File: internal/model/workflow.go
package model

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/tuanvumaihuynh/roboflow/pkg/paging"
	"github.com/tuanvumaihuynh/roboflow/pkg/xsort"
)

type Workflow struct {
	ID          uuid.UUID
	Name        string
	Description *string
	Definition  *WorkflowDefinition
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type WorkflowRepository interface {
	Get(ctx context.Context, id uuid.UUID) (Workflow, error)
	List(ctx context.Context, p paging.Params, sorts []xsort.Sort) (*paging.List[Workflow], error)
	Create(ctx context.Context, workflow Workflow) error
	Update(ctx context.Context, workflow Workflow) (Workflow, error)
	Delete(ctx context.Context, id uuid.UUID) error
}

File: ui/src/api/workflow.ts
import type { Paging } from '@/types/paging'
import type { Workflow, WorkflowDefinition, WorkflowWithoutDefinition } from '@/types/workflow'
import http from '@/lib/http'

export type ListWorkflowSort = 'name' | 'created_at' | 'updated_at'
export interface ListWorkflowParams {
  page?: number
  pageSize?: number
  sort?: Array<`-${ListWorkflowSort}` | ListWorkflowSort>
}

export interface CreateRaybotParams {
  name: string
  description?: string
  definition: WorkflowDefinition
}

export type UpdateRaybotParams = CreateRaybotParams

export interface RunWorkflowParams {
  env: {
    [key: string]: string
  }
}

const workflows = {
  list(p: ListWorkflowParams): Promise<Paging<WorkflowWithoutDefinition>> {
    return http.get('/workflows', { params: {
      page: p.page,
      pageSize: p.pageSize,
      sort: p.sort?.join(','),
    },
    })
  },
  get(id: string): Promise<Workflow> {
    return http.get(`/workflows/${id}`)
  },
  create(body: CreateRaybotParams): Promise<Workflow> {
    return http.post('/workflows', body)
  },
  update(id: string, body: UpdateRaybotParams): Promise<Workflow> {
    return http.put(`/workflows/${id}`, body)
  },
  delete(id: string): Promise<void> {
    return http.delete(`/workflows/${id}`)
  },
  run(id: string, body: RunWorkflowParams): Promise<void> {
    return http.post(`/workflows/${id}/run`, body)
  },
}

export default workflows
